File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/provider/AutoFillTest.java
package org.totschnig.myexpenses.test.provider;

import android.content.ContentUris;
import android.database.Cursor;

import org.totschnig.myexpenses.model.AccountType;
import org.totschnig.myexpenses.model.PaymentMethod;
import org.totschnig.myexpenses.provider.DatabaseConstants;
import org.totschnig.myexpenses.provider.TransactionProvider;
import org.totschnig.myexpenses.testutils.BaseDbTest;

import static org.totschnig.myexpenses.provider.DatabaseConstants.CAT_AS_LABEL;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ACCOUNTID;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_AMOUNT;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_CATID;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_COMMENT;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_CURRENCY;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_METHODID;

public class AutoFillTest extends BaseDbTest {

  // Contains the test data, as an array of TransactionInfo instances.
  private TransactionInfo1[] TEST_TRANSACTIONS = new TransactionInfo1[3];
  private String PAYEE_NAME = "N.N", PAYEE_NAME1 = "y.y";
  long testAccountId;
  long payeeId, payeeId1;
  long catId, catId1;
  long methodChequeId, methodCreditCardId;

  /**
   * Sets up test data.
   * The test data is in an SQL database. It is created in setUp() without any data,
   * and populated in insertData if necessary.
   */
  private void insertData() {

    long now = System.currentTimeMillis() / 1000;
    TEST_TRANSACTIONS[0] = new TransactionInfo1("Transaction 0", now, 0, testAccountId, payeeId, catId1, methodCreditCardId);
    TEST_TRANSACTIONS[1] = new TransactionInfo1("Transaction 1", now, 200, testAccountId, payeeId, catId, methodChequeId); //this is the transaction we should retrieve
    TEST_TRANSACTIONS[2] = new TransactionInfo1("Transaction 2", now, -100, testAccountId, payeeId1, catId, methodCreditCardId);

    // Sets up test data
    for (TransactionInfo1 TEST_TRANSACTION : TEST_TRANSACTIONS) {

      mDb.insertOrThrow(
          DatabaseConstants.TABLE_TRANSACTIONS,
          null,
          TEST_TRANSACTION.getContentValues()
      );
    }
  }

  @Override
  protected void setUp() throws Exception {
    super.setUp();
    AccountInfo testAccount = new AccountInfo("Test account", AccountType.CASH, 0, "USD");
    testAccountId = mDb.insertOrThrow(DatabaseConstants.TABLE_ACCOUNTS, null, testAccount.getContentValues());
    payeeId = mDb.insertOrThrow(DatabaseConstants.TABLE_PAYEES, null, new PayeeInfo(PAYEE_NAME).getContentValues());
    catId = mDb.insertOrThrow(DatabaseConstants.TABLE_CATEGORIES, null,  new CategoryInfo("Main", null).getContentValues());
    methodChequeId = PaymentMethod.find(PaymentMethod.PreDefined.CHEQUE.name());
    methodCreditCardId = PaymentMethod.find(PaymentMethod.PreDefined.CREDITCARD.name());

    payeeId1 = mDb.insertOrThrow(DatabaseConstants.TABLE_PAYEES, null, new PayeeInfo(PAYEE_NAME1).getContentValues());
    catId1 = mDb.insertOrThrow(DatabaseConstants.TABLE_CATEGORIES, null,  new CategoryInfo("Main 1", null).getContentValues());

    insertData();
  }

  public void testAutoLoadData() {
    final String[] projection = new String[]{KEY_CURRENCY, KEY_AMOUNT, KEY_CATID, CAT_AS_LABEL,
        KEY_COMMENT, KEY_ACCOUNTID, KEY_METHODID};
    Cursor cursor = getMockContentResolver().query(
        ContentUris.withAppendedId(TransactionProvider.AUTOFILL_URI, payeeId),
        projection,
        null,
        null,
        null);
    assert cursor != null;

    assertEquals(1, cursor.getCount());

    cursor.moveToFirst();

    assertEquals("USD", cursor.getString(0));
    assertEquals(200, cursor.getLong(1));
    assertEquals(catId, cursor.getLong(2));
    assertEquals("Main", cursor.getString(3));
    assertEquals("Transaction 1", cursor.getString(4));
    assertEquals(testAccountId, cursor.getLong(5));
    assertEquals(methodChequeId, cursor.getLong(6));

    cursor.close();
  }
}

File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/provider/AutoFillTest.kt
package org.totschnig.myexpenses.test.provider

import android.content.ContentUris
import org.totschnig.myexpenses.db2.findPaymentMethod
import org.totschnig.myexpenses.model.PreDefinedPaymentMethod
import org.totschnig.myexpenses.provider.DatabaseConstants
import org.totschnig.myexpenses.provider.DatabaseConstants.TABLE_CATEGORIES
import org.totschnig.myexpenses.provider.DatabaseConstants.TABLE_PAYEES
import org.totschnig.myexpenses.provider.PayeeInfo
import org.totschnig.myexpenses.provider.TransactionInfo
import org.totschnig.myexpenses.provider.TransactionProvider
import org.totschnig.myexpenses.provider.insert
import org.totschnig.myexpenses.testutils.BaseDbTest
import org.totschnig.myexpenses.testutils.CategoryInfo
import org.totschnig.shared_test.CursorSubject
import org.totschnig.shared_test.CursorSubject.Companion.useAndAssert

class AutoFillTest : BaseDbTest() {
    private val testTransactions = arrayOfNulls<TransactionInfo>(3)
    private var testAccountId: Long = 0
    private var payeeId: Long = 0
    private var payeeId1: Long = 0
    private var duplicateId: Long = 0
    private var payeeId2 : Long = 0
    private var catId: Long = 0
    private var catId1: Long = 0
    private var methodChequeId: Long = 0
    private var methodCreditCardId: Long = 0

    private fun insertData() {
        testTransactions[0] = TransactionInfo(
            accountId = testAccountId,
            amount = 0,
            comment = "Transaction 0",
            payeeId = payeeId,
            debtId = null,
            catId = catId1,
            methodId = methodCreditCardId
        )
        testTransactions[1] = TransactionInfo(
            accountId = testAccountId,
            amount = 200,
            comment = "Transaction 1",
            payeeId = payeeId,
            debtId = null,
            catId = catId,
            methodId = methodChequeId
        )
        testTransactions[2] = TransactionInfo(
            accountId = testAccountId,
            amount = -100,
            comment = "Transaction 2",
            payeeId = payeeId1,
            debtId = null,
            catId = catId,
            methodId = methodCreditCardId
        )

        for (testTransaction in testTransactions) {
            mDb.insert(DatabaseConstants.TABLE_TRANSACTIONS, testTransaction!!.contentValues)
        }
    }

    @Deprecated("Deprecated in Java")
    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()
        testAccountId = setupTestAccount()
        payeeId = mDb.insert(TABLE_PAYEES, PayeeInfo("N.N").contentValues)
        catId =
            mDb.insert(TABLE_CATEGORIES, CategoryInfo("Main").contentValues)
        methodChequeId = repository.findPaymentMethod(PreDefinedPaymentMethod.CHEQUE.name)!!
        methodCreditCardId = repository.findPaymentMethod(PreDefinedPaymentMethod.CREDITCARD.name)!!
        payeeId1 = mDb.insert(TABLE_PAYEES, PayeeInfo("y.y").contentValues)
        duplicateId = mDb.insert(TABLE_PAYEES, PayeeInfo("yy", payeeId1).contentValues)
        payeeId2 = mDb.insert(TABLE_PAYEES, PayeeInfo("OhneGeld").contentValues)
        catId1 = mDb.insert(
            TABLE_CATEGORIES,
            CategoryInfo("Main 1").contentValues
        )
        insertData()
    }

    fun testAutoLoadData() {
        testAutoLoadForPayee(payeeId, testTransactions[1]!!)
    }

    fun testAutoLoadDataForDuplicate() {
        testAutoLoadForPayee(duplicateId, testTransactions[2]!!)
    }

    fun testAutoLoadDataForPayeeWithoutTransaction() {
        testAutoLoadForPayee(payeeId2, null)
    }

    private fun testAutoLoadForPayee(payeeId: Long, transaction: TransactionInfo?) {
        val projection = arrayOf(
            DatabaseConstants.KEY_CURRENCY,
            DatabaseConstants.KEY_AMOUNT,
            DatabaseConstants.KEY_CATID,
            DatabaseConstants.CAT_AS_LABEL,
            DatabaseConstants.KEY_COMMENT,
            DatabaseConstants.KEY_ACCOUNTID,
            DatabaseConstants.KEY_METHODID
        )
        mockContentResolver.query(
            ContentUris.withAppendedId(TransactionProvider.AUTOFILL_URI, payeeId),
            projection,
            null,
            null,
            null
        ).useAndAssert {
            if (transaction != null) {
                reportsTransaction(transaction)
            } else {
                hasCount(0)
            }
        }
    }

    private fun CursorSubject.reportsTransaction(transaction: TransactionInfo) {
        hasCount(1)
        movesToFirst()
        hasLong(1, transaction.amount)
        hasLong(2, transaction.catId!!)
        hasString(3, "Main")
        hasString(4, transaction.comment)
        hasLong(5, transaction.accountId)
        hasLong(6, transaction.methodId!!)
    }
}
